File: 53.php
<?php

class Solution {

    /**
     * @param Integer[] $nums
     * @return Integer
     */
    function maxSubArray($nums) {
        $length = count($nums);
        $dp0 = $nums[$length - 1];
        $max = $nums[$length - 1];

        for ($i = $length - 2; $i >= 0; $i--) {
            $dp1 = max($nums[$i], $dp0 + $nums[$i]);
            $dp0 = $dp1;
            $max = max($max, $dp1);
        }
        return $max;
    }
}

File: go/dp/53.go
package main

func main() {

}

func maxSubArray(nums []int) int {
	dp, maxNum := nums[0], nums[0]

	for i := 1; i < len(nums); i++ {
		dp = GetMax(nums[i], dp+nums[i])
		maxNum = GetMax(maxNum, dp)
	}
	return maxNum
}

func GetMax(x int, y int) int {
	if x > y {
		return x
	}
	return y
}